design/StableMemory.md

```diff
@@ -0,0 +1,290 @@
+stable memory, some coordination between these two alternative, co-existing
+interfaces to IC stable memory is required.
+
+
+# The IC's Stable Memory API
+
+The IC provides a very small set of functions for operating on stable memory:
+
+```
+ic0.stable_size : () -> (page_count : i32);                                 // *
+ic0.stable_grow : (new_pages : i32) -> (old_page_count : i32);              // *
+ic0.stable_write : (offset : i32, src : i32, size : i32) -> ();             // *
+ic0.stable_read : (dst : i32, offset : i32, size : i32) -> ();              // *
+```
+
+(see https://sdk.dfinity.org/docs/interface-spec/index.html#system-api-stable-memory)
+
+These grow memory and do bulk transfers between Wasm and stable
+memory.  The `// *` means that they can be called in all contexts
+(e.g. init, update, query etc).  Direct reads and writes of word-sized
+data to/from the stack are not supported but can be emulated at cost.
+The contents of fresh pages (after grow) is initially zero.
+
+Note that, in this API, the client is responsible for growing (both
+stable and wasm) memory before access by read or write (out-of-bounds
+access will trap).
+
+# A minimal Stable Memory API
+
+The minimal Motoko prims could be:
+
+```
+module StableMemory {
+  size : () -> (logical_page_count : i32); // <= ic0.stable_size()
+  grow : () -> (new_pages : i32) -> (old_logical_page_count : i32);
+  loadNat8 : (offset : Nat32) -> Nat8;
+    // traps outside logical address space
+  storeNat8 : (offset : Nat32, n : Nat8) -> ();
+    // traps outside logical address space
+  ...
+  loadBlob : (offset : Nat32, size : i32) -> Blob
+     // read Blob contents from memory at [offset,..,offset+size-1] into fresh blob, trapping if exceeding logical address space
+  storeBlob : (offset : Nat32, b : Blob) -> (); // write contents of blob to memory, trapping if exceeding logical address space
+}
+```
+
+NOTE: Motoko's `Nat32` value are always boxed - it might be more efficient to use `Nat` which is unboxed for 30(?)-bit values.
+
+
+```
+fun loadNat8(offset, b) =
+   assert (offset < StableMemory.size() * wasm_page_size);
+   mem[offset]
+
+fun storeNat8(offset, b) =
+   assert (offset < StableMemory.size() * wasm_page_size);
+   mem[offset] := b
+
+```
+
+(To avoid overflow on the rhs, we could implement the check as `assert ((offset >> 6) < StableMemory.size())`.)
+
+On top of this basic API, users should be able to build more
+interesting higher-level APIs for pickling user-defined data.
+
+REMARK:
+
+Actually implementing the sketched assignments using the existing IC
+API involves writing the contents to memory and then copying stable
+memory - even for individual words - this could be optimized by an
+improved system API offering direct load and stores from/to the stack:
+
+```
+ic0.stable_write_i32 : (offset : i32, val: i32) -> ();   // *
+ic0.stable_read_i32 : (offset : i32, size : i32) -> i32; // *
+// similarly for i64, f32, f64
```

Comment:
You'll likely also want i8 and i16 versions (that extend to i32 values, either signed or unsigned). Essentially, you want the analogous of all of Wasm's load/store instructions.